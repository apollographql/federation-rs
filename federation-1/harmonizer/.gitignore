/target
/dist
/node_modules

js-dist/
bundled/
tsconfig.tsbuildinfo