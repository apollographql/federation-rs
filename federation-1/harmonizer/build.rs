use deno_core::{JsRuntimeForSnapshot, RuntimeOptions};
use semver::Version;
use serde_json::Value as JsonValue;
use std::path::PathBuf;
use std::{env, error::Error, fs, io::Write, path::Path, process::Command};
use toml_edit::{value as new_toml_value, Document as TomlDocument};

// this build.rs file is used by both `federation-1/harmonizer` and `federation-2/harmonizer`
// to keep the crate version in line with the appropriate npm package
// and to build the V8 snapshots

fn main() {
    // Always rerun the script
    let out_dir = std::env::var_os("OUT_DIR").expect("$OUT_DIR not set.");
    println!("cargo:rerun-if-changed={:?}", &out_dir);
    let out_dir: PathBuf = out_dir.into();
    if cfg!(target_arch = "musl") {
        panic!("This package cannot be built for musl architectures.");
    }

    let current_dir = std::env::current_dir().unwrap();

    // only do `npm` related stuff if we're _not_ publishing to crates.io
    // package.json is not in the `includes` section of `Cargo.toml`
    if std::fs::metadata("./package.json").is_ok() {
        update_manifests();
        bundle_for_deno(&current_dir);
    }

    // always create the snapshot
    create_snapshot(&out_dir).expect("unable to create v8 snapshot: composition.snap");
}

// runs `npm install` && `npm run build` in the current `harmonizer-x` workspace crate
fn bundle_for_deno(current_dir: &Path) {
    let npm = which::which("npm").expect("You must have npm installed to build this crate.");

    if cfg!(debug_assertions) {
        // in debug mode we want to update the package-lock.json
        // so we run `npm install`
        println!(
            "cargo:warning=running `npm install` in {}",
            &current_dir.display()
        );
        assert!(Command::new(&npm)
            .current_dir(current_dir)
            .args(["install"])
            .status()
            .expect("Could not get status of `npm install`")
            .success());
    } else {
        // in release mode, we're probably running in CI
        // and want the version we publish to match
        // the git source
        // so we run `npm ci`.
        println!(
            "cargo:warning=running `npm ci` in {}",
            &current_dir.display()
        );
        assert!(Command::new(&npm)
            .current_dir(current_dir)
            .args(["ci"])
            .status()
            .expect("Could not get status of `npm ci`")
            .success());
    }

    println!(
        "cargo:warning=running `npm run format` in {}",
        &current_dir.display()
    );
    assert!(Command::new(&npm)
        .current_dir(current_dir)
        .args(["run", "format"])
        .status()
        .expect("Could not get status of `npm run format`")
        .success());

    println!(
        "cargo:warning=running `npm run build` in {}",
        &current_dir.display()
    );
    assert!(Command::new(&npm)
        .current_dir(current_dir)
        .args(["run", "build"])
        .status()
        .expect("Could not get status of `npm run build`")
        .success());
}

// updates `Cargo.toml` and `package.json` in the current `federation-x/harmonizer` crate
fn update_manifests() {
    let current_dir = std::env::current_dir().expect("Could not find the current directory.");
    let harmonizer_manifest_path = current_dir.join("Cargo.toml");
    let maybe_harmonizer_version = update_this_manifest(&harmonizer_manifest_path);
    if let Some(harmonizer_version) = maybe_harmonizer_version {
        println!(
            "cargo:warning=updated {} to {}",
            &harmonizer_manifest_path.display(),
            &harmonizer_version
        );
        let federation_workspace_dir = current_dir
            .parent()
            .expect("Could not find parent directory.");
        let supergraph_dir = federation_workspace_dir.join("supergraph");
        let supergraph_manifest_path = supergraph_dir.join("Cargo.toml");
        update_supergraph_manifest(&supergraph_manifest_path, &harmonizer_version);
    }
}

// Updates the `Cargo.toml` for this version of harmonizer
// and returns Some(Version) if it was updated and None if it was not
fn update_this_manifest(build_manifest_path: &Path) -> Option<Version> {
    let build_manifest_contents =
        fs::read_to_string(build_manifest_path).expect("Could not read 'Cargo.toml'");
    let mut build_manifest = build_manifest_contents
        .parse::<TomlDocument>()
        .expect("Cargo.toml is not valid TOML");

    let js_composition_version = get_underlying_composition_npm_module_version();

    let crate_version = Version::parse(
        build_manifest["package"]["version"]
            .as_str()
            .expect("`package.version` in Cargo.toml is not a string"),
    )
    .expect("Crate version is not valid semver");

    if js_composition_version != crate_version {
        build_manifest["package"]["version"] = new_toml_value(js_composition_version.to_string());
        fs::write(build_manifest_path, build_manifest.to_string())
            .expect("Could not write updated Cargo.toml");
        Some(js_composition_version)
    } else {
        None
    }
}

fn update_supergraph_manifest(supergraph_manifest_path: &Path, new_package_version: &Version) {
    let supergraph_manifest_contents =
        fs::read_to_string(supergraph_manifest_path).expect("Could not read Cargo.toml");
    let mut supergraph_manifest = supergraph_manifest_contents
        .parse::<TomlDocument>()
        .expect("Cargo.toml is not valid TOML");
    supergraph_manifest["package"]["version"] = new_toml_value(new_package_version.to_string());
    fs::write(supergraph_manifest_path, supergraph_manifest.to_string())
        .expect("Could not update Cargo.toml");
}

// reads package.json, finds the correct composition JS dependency, and returns its version
fn get_underlying_composition_npm_module_version() -> Version {
    let current_dir = env::current_dir().unwrap();
    let npm_manifest_path = current_dir.join("package.json");
    let mut npm_manifest_contents: JsonValue = serde_json::from_str(
        &fs::read_to_string(&npm_manifest_path).expect("Could not read package.json"),
    )
    .expect("package.json is not valid JSON");

    let maybe_federation = npm_manifest_contents["dependencies"]["@apollo/federation"].as_str();
    let maybe_composition = npm_manifest_contents["dependencies"]["@apollo/composition"].as_str();
    let (dep_name, version_string) = match (maybe_federation, maybe_composition) {
        (None, Some(composition)) => {
            let dep_name = "@apollo/composition".to_string();
            let version_str = npm_manifest_contents["dependencies"][&dep_name]
                .as_str()
                .unwrap_or_else(|| panic!("`.dependencies.{}` is not a string", &composition));
            (dep_name, version_str.to_string())
        }
        (Some(federation), None) => {
            let dep_name = "@apollo/federation".to_string();
            let version_str = npm_manifest_contents["dependencies"][&dep_name]
                .as_str()
                .unwrap_or_else(|| panic!("`.dependencies.{}` is not a string", &federation));
            (dep_name, version_str.to_string())
        }
        (Some(_federation), Some(_composition)) => unreachable!(
            "Found both `@apollo/federation` and `@apollo/composition`. There should only be one."
        ),
        (None, None) => unreachable!(
            "Underlying npm module must be either `@apollo/federation` or `@apollo/composition`"
        ),
    };

    let parsed_version = Version::parse(&version_string).unwrap_or_else(|_| {
        panic!(
            "version for `{}`, `{}`, is not valid semver",
            &dep_name, &version_string
        )
    });

    npm_manifest_contents["version"] = JsonValue::from(version_string);
    fs::write(
        npm_manifest_path,
        serde_json::to_string_pretty(&npm_manifest_contents).expect("Could not pretty print JSON"),
    )
    .expect("Could not write updated contents to package.json");

    parsed_version
}

fn create_snapshot(out_dir: &Path) -> Result<(), Box<dyn Error>> {
    let options = RuntimeOptions {
        ..Default::default()
    };
    let mut runtime = JsRuntimeForSnapshot::new(options);

    // The runtime automatically contains a Deno.core object with several
    // functions for interacting with it.
    let runtime_source = fs::read_to_string("bundled/runtime.js")?;
    runtime
        .execute_script(
            "<init>",
            deno_core::FastString::Owned(runtime_source.into()),
        )
        .expect("unable to initialize harmonizer runtime environment");

    // Load the composition library.
    let composition_source = fs::read_to_string("bundled/composition_bridge.js")?;
    runtime
        .execute_script(
            "composition_bridge.js",
            deno_core::FastString::Owned(composition_source.into()),
        )
        .expect("unable to evaluate composition module");

    // Create our base query snapshot which will be included in
    // src/js.rs to initialise our JsRuntime().
    println!("cargo:warning={:?}", &out_dir);
    let mut snap = fs::File::create(out_dir.join("composition.snap"))?;
    snap.write_all(&runtime.snapshot())?;

    Ok(())
}
