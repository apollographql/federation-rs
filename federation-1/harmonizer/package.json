{
  "name": "@apollo/harmonizer-0",
  "private": true,
  "version": "0.37.1",
  "description": "Apollo Federation Harmonizer JS Entrypoint",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "make-dir bundled js-dist && rm -f tsconfig.tsbuildinfo && tsc --build --verbose && node esbuild/bundler.js && cp js-dist/runtime.js js-dist/do_compose.js bundled/",
    "clean": "rm -rf ./node_modules ./js-dist ./bundled ./tsconfig.tsbuildinfo",
    "lint": "prettier --check ./esbuild/**/*.js ./js-src/**/*.ts ./js-src/**/*.js",
    "format": "prettier --write ./esbuild/**/*.js ./js-src/**/*.ts ./js-src/**/*.js"
  },
  "browser": {
    "buffer": "buffer",
    "crypto": "crypto-js",
    "util": "util"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/apollographql/federation.git",
    "directory": "harmonizer-js/"
  },
  "author": "Apollo <packages@apollographql.com>",
  "license": "MIT",
  "engines": {
    "node": ">=12.13.0 <17.0",
    "npm": ">=7 <9"
  },
  "dependencies": {
    "@apollo/federation": "0.37.1"
  },
  "peerDependencies": {
    "graphql": "^14.5.0 || ^15.0.0"
  },
  "volta": {
    "node": "16.13.2",
    "npm": "8.3.1"
  },
  "devDependencies": {
    "buffer": "^6.0.3",
    "crypto-js": "^4.2.0",
    "esbuild": "^0.14.14",
    "make-dir-cli": "^3.0.0",
    "prettier": "^2.5.1",
    "typescript": "^5.0.2",
    "util": "^0.12.4"
  }
}