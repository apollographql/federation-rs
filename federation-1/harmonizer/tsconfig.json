{
  "compilerOptions": {
    "composite": true,
    "target": "es2019",
    "module": "commonjs",
    "moduleResolution": "node",
    "esModuleInterop": true,
    "sourceMap": true,
    "declaration": true,
    "declarationMap": true,
    "removeComments": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "noUnusedParameters": true,
    "noUnusedLocals": true,
    "useUnknownInCatchVariables": false,
    "forceConsistentCasingInFileNames": true,
    "lib": ["es2021", "esnext.asynciterable"],
    "rootDir": "./js-src",
    "outDir": "./js-dist",
    "types": [],
    "allowJs": true,
    "strict": false,
  },
  "include": ["./js-src/**/*"],
  "exclude": ["**/__tests__"]
}
