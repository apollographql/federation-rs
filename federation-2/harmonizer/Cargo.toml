[package]
name = "harmonizer"
version = "2.0.2"
authors = ["Apollo <packages@apollographql.com>"]
edition = "2018"
description = "Apollo Federation utility to compose a supergraph from subgraphs"
homepage = "https://www.apollographql.com/apollo-federation/"
documentation = "https://apollographql.com/docs/federation/"
repository = "https://github.com/apollographql/federation/"
license-file = "./LICENSE"
readme = "README.md"
include = [
  "dist/**/*",
  "src/**/*",
  "deno/**/*",
  "Cargo.toml",
  "build.rs",
]

[dependencies]
apollo-federation-types = { version = "0.5", path = "../../apollo-federation-types", default-features = false, features = [
  "build",
] }
deno_core = "0.118.0"
serde = { version = "1", features = ["derive"] }
serde_json = "1"

[dev-dependencies]
insta = "1.8.0"

[build-dependencies]
deno_core = "0.118.0"
semver = "1"
serde_json = "1"
toml_edit = "0.13"
which = "4.2.2"
