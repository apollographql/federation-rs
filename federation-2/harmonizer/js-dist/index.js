Object.defineProperty(exports, "__esModule", { value: true });
exports.composition = void 0;
var composition_1 = require("./composition");
Object.defineProperty(exports, "composition", { enumerable: true, get: function () { return composition_1.composition; } });
//# sourceMappingURL=index.js.map