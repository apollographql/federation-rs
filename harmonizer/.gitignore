/target
Cargo.lock
/dist
/node_modules