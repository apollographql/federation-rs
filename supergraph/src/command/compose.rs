use camino::Utf8PathBuf;
use structopt::StructOpt;

use apollo_federation_types::{
    build::BuildResult,
    config::{ConfigError, PluginVersion, SupergraphConfig},
};
use harmonizer::harmonize;

use std::fs::File;
use std::io::BufWriter;

#[derive(Debug, StructOpt)]
pub struct Compose {
    /// The path to the fully resolved supergraph YAML.
    ///
    /// NOTE: Each subgraph entry MUST contain raw SDL
    /// as the schema source.
    config_file: Utf8PathBuf,
    /// Output to a file. Default is to use stdout.
    output: Option<Utf8PathBuf>,
}

impl Compose {
    pub fn run(&self) -> ! {
        let composition_result = self.do_compose();
        let composition_succeeded = composition_result.is_ok();

        if let Some(filepath) = &self.output {
            let json = composition_result;
            let file =
                File::create(filepath).expect(&format!("failed to create file at ${filepath}"));
            let writer = BufWriter::new(file);
            serde_json::to_writer(writer, &json)
                .expect("failed to write composition result to file");
        } else {
            print!("{}", serde_json::json!(composition_result));
        }

        if composition_succeeded {
            std::process::exit(0);
        } else {
            std::process::exit(1);
        }
    }

    fn do_compose(&self) -> BuildResult {
        let supergraph_config = SupergraphConfig::new_from_yaml_file(&self.config_file)?;
        if let Some(federation_version) = supergraph_config.get_federation_version() {
            if !matches!(federation_version.get_major_version(), 2) {
                return Err(ConfigError::InvalidConfiguration {message: format!("It looks like '{}' resolved to 'federation_version: {}', which doesn't match the current supergraph binary.", &self.config_file, federation_version )}.into());
            }
        }
        let subgraph_definitions = supergraph_config.get_subgraph_definitions()?;
        harmonize(subgraph_definitions)
    }
}

#[test]
fn compose_test() {
    let res = Compose {
        config_file: "./tests/compose_test.yaml".into(),
        output: None,
    };

    let result = res.do_compose();

    insta::assert_json_snapshot!(result);
}
