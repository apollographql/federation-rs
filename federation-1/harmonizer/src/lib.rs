/*!
# Harmonizer

This _harmonizer_ offers the ability to invoke a bundled version of the
JavaScript library, [`@apollo/federation`], which _composes_ multiple subgraphs
into a supergraph.

The bundled version of the federation library that is included is a JavaScript
Immediately Invoked Function Expression ([IIFE]) that is created by running the
[esbuild] bundler on the `@apollo/federation` package.

When the [`harmonize`] function that this crate provides is called with a
[`ServiceList`] (which is synonymous with the terminology and service list
notion that exists within the JavaScript composition library), this crate uses
[`deno_core`] to invoke the JavaScript within V8.  This is ultimately
accomplished using [`rusty_v8`]'s V8 bindings to V8.

While we intend for a future version of composition to be done natively within
Rust, this allows us to provide a more stable transition using an already stable
composition implementation while we work toward something else.

[`@apollo/federation`]: https://npm.im/@apollo/federation
[IIFE]: https://developer.mozilla.org/en-US/docs/Glossary/IIFE
[esbuild]: http://esbuild.github.io/
[`deno_core`]: https://crates.io/crates/deno_core
[`rusty_v8`]: https://crates.io/crates/rusty_v8
*/

#![forbid(unsafe_code)]
#![deny(missing_debug_implementations, nonstandard_style)]
#![warn(missing_docs, future_incompatible, unreachable_pub, rust_2018_idioms)]
use deno_core::{error::AnyError, op, Extension, JsRuntime, OpState, RuntimeOptions, Snapshot};
use std::sync::mpsc::{channel, Sender};

mod js_types;

use js_types::CompositionError;

use apollo_federation_types::build::{
    BuildError, BuildErrors, BuildOutput, BuildResult, SubgraphDefinition,
};

/// The `harmonize` function receives a [`Vec<SubgraphDefinition>`] and invokes JavaScript
/// composition on it, either returning the successful output, or a list of error messages.
pub fn harmonize(subgraph_definitions: Vec<SubgraphDefinition>) -> BuildResult {
    // The snapshot is created in the build_harmonizer.rs script and included in our binary image
    let buffer = include_bytes!(concat!(env!("OUT_DIR"), "/composition.snap"));

    // We'll use this channel to get the results
    let (tx, rx) = channel::<Result<BuildOutput, BuildErrors>>();

    let my_ext = Extension::builder("harmonizer")
        .ops(vec![op_composition_result::decl()])
        .state(move |state| {
            state.put(tx.clone());
            Ok(())
        })
        .build();

    // Use our snapshot to provision our new runtime
    let options = RuntimeOptions {
        startup_snapshot: Some(Snapshot::Static(buffer)),
        extensions: vec![my_ext],
        ..Default::default()
    };
    let mut runtime = JsRuntime::new(options);

    // convert the subgraph definitions into JSON
    let service_list_javascript = format!(
        "serviceList = {}",
        serde_json::to_string(&subgraph_definitions)
            .expect("unable to serialize service list into JavaScript runtime")
    );

    // store the subgraph definition JSON in the `serviceList` variable
    runtime
        .execute_script("<set_service_list>", &service_list_javascript)
        .expect("unable to evaluate service list in JavaScript runtime");

    // run the unmodified do_compose.js file, which expects `serviceList` to be set
    runtime
        .execute_script("do_compose.js", include_str!("../bundled/do_compose.js"))
        .expect("unable to invoke composition in JavaScript runtime");

    // wait for a message from `op_composition_result`
    rx.recv().expect("channel remains open")
}

#[op]
fn op_composition_result(
    state: &mut OpState,
    value: serde_json::Value,
) -> Result<serde_json::Value, AnyError> {
    // the JavaScript object can contain an array of errors
    let deserialized_result: Result<Result<String, Vec<CompositionError>>, serde_json::Error> =
        serde_json::from_value(value);

    let build_result: Result<String, Vec<CompositionError>> = match deserialized_result {
        Ok(build_result) => build_result,
        Err(e) => Err(vec![CompositionError::generic(format!(
            "Something went wrong, this is a bug: {e}"
        ))]),
    };

    let build_output = build_result
        .map(BuildOutput::new)
        .map_err(|composition_errors| {
            // we then embed that array of errors into the `BuildErrors` type which is implemented
            // as a single error with each of the underlying errors listed as causes.
            composition_errors
                .iter()
                .map(|err| BuildError::from(err.clone()))
                .collect::<BuildErrors>()
        });

    let sender = state
        .borrow::<Sender<Result<BuildOutput, BuildErrors>>>()
        .clone();
    // send the build result
    sender.send(build_output).expect("channel must be open");

    // Don't return anything to JS since its value is unused
    Ok(serde_json::json!(null))
}

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        use crate::{harmonize, SubgraphDefinition};

        insta::assert_snapshot!(
            harmonize(vec![
                SubgraphDefinition::new(
                    "users",
                    "undefined",
                    "
            type User @key(fields: \"id\") {
              id: ID
              name: String
            }

            extend type Query {
              users: [User!]
            }
          "
                ),
                SubgraphDefinition::new(
                    "movies",
                    "undefined",
                    "
            type Movie {
              title: String
              name: String
            }

            extend type User {
              favorites: [Movie!]
            }

            extend type Query {
              movies: [Movie!]
            }
          "
                )
            ])
            .unwrap()
            .supergraph_sdl
        );
    }
}
