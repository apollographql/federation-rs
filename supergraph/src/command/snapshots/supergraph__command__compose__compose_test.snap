---
source: supergraph/src/command/compose.rs
expression: result
---
{
  "Ok": {
    "supergraphSdl": "schema\n  @link(url: \"https://specs.apollo.dev/link/v1.0\")\n  @link(url: \"https://specs.apollo.dev/join/v0.4\", for: EXECUTION)\n  @join__directive(graphs: [TEST], name: \"link\", args: {url: \"https://specs.apollo.dev/source/v0.1\", import: [\"@sourceAPI\", \"@sourceType\", \"@sourceField\"], for: \"EXECUTION\"})\n  @join__directive(graphs: [TEST], name: \"sourceAPI\", args: {name: \"json\", http: {baseURL: \"http://a.co/\"}})\n{\n  query: Query\n}\n\ndirective @join__directive(graphs: [join__Graph!], name: String!, args: join__DirectiveArguments) repeatable on SCHEMA | OBJECT | INTERFACE | FIELD_DEFINITION\n\ndirective @join__enumValue(graph: join__Graph!) repeatable on ENUM_VALUE\n\ndirective @join__field(graph: join__Graph, requires: join__FieldSet, provides: join__FieldSet, type: String, external: Boolean, override: String, usedOverridden: Boolean, overrideLabel: String) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION\n\ndirective @join__graph(name: String!, url: String!) on ENUM_VALUE\n\ndirective @join__implements(graph: join__Graph!, interface: String!) repeatable on OBJECT | INTERFACE\n\ndirective @join__type(graph: join__Graph!, key: join__FieldSet, extension: Boolean! = false, resolvable: Boolean! = true, isInterfaceObject: Boolean! = false) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR\n\ndirective @join__unionMember(graph: join__Graph!, member: String!) repeatable on UNION\n\ndirective @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA\n\ninput HTTPHeaderMapping\n  @join__type(graph: TEST)\n{\n  name: String!\n  as: String\n  value: String\n}\n\ninput HTTPSourceAPI\n  @join__type(graph: TEST)\n{\n  baseURL: String!\n  headers: [HTTPHeaderMapping!]\n}\n\ninput HTTPSourceField\n  @join__type(graph: TEST)\n{\n  GET: source__URLPathTemplate\n  POST: source__URLPathTemplate\n  PUT: source__URLPathTemplate\n  PATCH: source__URLPathTemplate\n  DELETE: source__URLPathTemplate\n  body: source__JSONSelection\n  headers: [HTTPHeaderMapping!]\n}\n\ninput HTTPSourceType\n  @join__type(graph: TEST)\n{\n  GET: source__URLPathTemplate\n  POST: source__URLPathTemplate\n  headers: [HTTPHeaderMapping!]\n  body: source__JSONSelection\n}\n\nscalar join__DirectiveArguments\n\nscalar join__FieldSet\n\nenum join__Graph {\n  TEST @join__graph(name: \"test\", url: \"http://localhost:4000/graphql\")\n}\n\ninput KeyTypeMap\n  @join__type(graph: TEST)\n{\n  key: String!\n  typeMap: source__TypenameKeyMap\n}\n\nscalar link__Import\n\nenum link__Purpose {\n  \"\"\"\n  `SECURITY` features provide metadata necessary to securely resolve fields.\n  \"\"\"\n  SECURITY\n\n  \"\"\"\n  `EXECUTION` features provide metadata necessary for operation execution.\n  \"\"\"\n  EXECUTION\n}\n\ntype Query\n  @join__type(graph: TEST)\n{\n  users: [User] @join__directive(graphs: [TEST], name: \"sourceField\", args: {api: \"json\", http: {GET: \"/users\"}, selection: \"id name\"})\n}\n\nscalar source__JSONSelection\n  @join__type(graph: TEST)\n\nscalar source__TypenameKeyMap\n  @join__type(graph: TEST)\n\nscalar source__URLPathTemplate\n  @join__type(graph: TEST)\n\ntype User\n  @join__type(graph: TEST)\n{\n  id: ID!\n  name: String\n}",
    "hints": []
  }
}
